src/cmd/flux-account-service.py

```diff
@@ -0,0 +1,467 @@
+from fluxacct.accounting import job_archive_interface as jobs
+from fluxacct.accounting import queue_subcommands as qu
+from fluxacct.accounting import project_subcommands as p
+
+
+def establish_sqlite_connection(path):
+    # try to open database file; will exit with -1 if database file not found
+    if not os.path.isfile(path):
+        print(f"Database file does not exist: {path}", file=sys.stderr)
+        sys.exit(1)
+
+    db_uri = "file:" + path + "?mode=rw"
+    try:
+        conn = sqlite3.connect(db_uri, uri=True)
+        # set foreign keys constraint
+        conn.execute("PRAGMA foreign_keys = 1")
+    except sqlite3.OperationalError as exc:
+        print(f"Unable to open database file: {db_uri}", file=sys.stderr)
+        print(f"Exception: {exc}")
+        sys.exit(1)
+
+    return conn
+
+
+def background():
+    pid = os.fork()
+    if pid > 0:
+        # exit first parent
+        sys.exit(0)
+
+
+def check_db_version(conn):
+    # check version of database; if not up to date, output message
+    # and exit
+    cur = conn.cursor()
+    cur.execute("PRAGMA user_version")
+    db_version = cur.fetchone()[0]
+    if db_version < 20:
+        print(
+            "flux-accounting database out of date; please update DB with 'flux account-update-db' before running commands"
+        )
+        sys.exit(1)
+
+
+# pylint: disable=broad-except
+class AccountingService:
+    def __init__(self, flux_handle, conn):
+
+        self.handle = flux_handle
+        self.conn = conn
+
+        try:
+            # register service with broker
+            self.handle.service_register("accounting").get()
+            print("registered accounting service", file=sys.stderr)
+        except FileExistsError:
+            LOGGER.error("flux-accounting service is already registered")
+
+        # register signal watcher for SIGTERM to initiate shutdown
+        self.handle.signal_watcher_create(signal.SIGTERM, self.shutdown).start()
+        self.handle.signal_watcher_create(signal.SIGINT, self.shutdown).start()
+
+        endpoints = [
+            "view_user",
+            "add_user",
+            "delete_user",
+            "edit_user",
+            "view_bank",
+            "add_bank",
+            "delete_bank",
+            "edit_bank",
+            "view_job_records",
+            "update_usage",
+            "add_queue",
+            "view_queue",
+            "delete_queue",
+            "edit_queue",
+            "add_project",
+            "view_project",
+            "delete_project",
+            "shutdown",
+        ]
+
+        for name in endpoints:
+            self.handle.msg_watcher_create(
+                getattr(self, name), FLUX_MSGTYPE_REQUEST, f"accounting.{name}", self
+            ).start()
+
+    def shutdown(self, handle, watcher, signum, arg):
+        print("Shutting down...", file=sys.stderr)
+        self.conn.close()
+        self.handle.service_unregister("accounting").get()
+        self.handle.reactor_stop()
```

Comment:
Oops! Yeah, while I was looking at this I noticed that I wasn't shutting down the service in the tests either, so I wouldn't have caught this. Thanks. I just force-pushed a fix to add a `handle.respond(signum)` to the `"shutdown"` message as well as a test in all of the sharness tests that use the service to call `"shutdown"` at the end of the test file.

Reply:
Just realized that I don't think that `handle.respond(signum)` is the right way to respond to this message, as when I go to shut down the service, there is some error messages that pop up:

```console
flux-uri: ERROR: 
Invalid arguments passed to wrapped C function:
Name: flux_respond
C signature: int(*)(struct flux_handle_struct *, struct flux_msg *, char *)
Arguments: (15, <cdata 'void *' NULL>)
          

[1]+  Exit 1                  flux account-service -p /tmp/test.db
```

Do you know of the right way to respond to this message? The parameters for this function `(self, handle, watcher, signum, arg)` are different from the other functions `(self, handle, watcher, msg, arg)`.

Reply:
Ah, you are using the same callback for signal watcher and the `shutdown` message handler, so the arguments are not correct. You'll need a separate handler for the message watcher, with the same parameters as the other message watcher, then use `handle.respond(msg)` (which sends a response with an empty payload)

Reply:
Oh okay! Thanks, I think that makes sense now. In this case, the signal watcher does not need to respond since it does not take a message from an RPC, correct? Its shutdown process can just end with `self.handle.reactor_stop()`?

Reply:
Nonetheless, I just force-pushed a fix to correct the signal watcher and added a `shutdown_service` message handler (the latter actually responds with `handle.respond(msg)`.

Reply:
Yeah, the signal watcher has nothing to respond to, so no response needed there.